Trusted types attributes

lukewarlow opened the pull request:
This calls the get Trusted Types-compliant attribute value algorithm from Trusted Types (https://github.com/w3c/trusted-types/pull/418) from attribute's change, append, and replace.

Changed the signature of setAttribute and setAttributeNS to accept Trusted Types as values. The underlying Attr node's values continue to be DOMString, so moving nodes across elements or adding standalone attributes to elements can cause TT violations. This matches WPT tests and the Chromium's implementation.

See and https://github.com/whatwg/dom/issues/789. Supercedes PR https://github.com/whatwg/dom/pull/809.

- [x] At least two implementers are interested (and none opposed):
   * Chromium
   * Gecko
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/tree/master/trusted-types
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium:https://bugs.chromium.org/p/chromium/issues/detail?id=739170
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1508286
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=266630
- [ ] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: …
- [ ] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

lukewarlow commented:
This is a clone of #1247 where I will finish any outstanding work to get this across the line

lukewarlow commented:
See also https://github.com/whatwg/dom/pull/1258 which is another integration point with the DOM spec that we need for TT.

otherdaniel commented on dom.bs:
```diff
@@ -6484,8 +6514,19 @@ string <var>namespace</var> (default null):</p>
- <li><p>If <var>oldAttr</var> is non-null, then <a lt="replace an attribute">replace</a>
- <var>oldAttr</var> with <var>attr</var>.
+ <li>
+  <p>If <var>oldAttr</var> is non-null, then:</p>
```
Would this condition result in the TT-check being skipped if the old attribute didn't exist?

lukewarlow replied:
In that case append an attribute is called which runs validate and set as it's first step, so TT check still happens

lukewarlow replied:
I need to check if replace is used elsewhere, if it's not I might move the validate and set call into it, to be consistent with other places.

lukewarlow replied:
I don't think replace is used anywhere else so will move the validation step a level down.

lukewarlow commented on dom.bs:
```diff
@@ -6484,8 +6514,19 @@ string <var>namespace</var> (default null):</p>
+   <li><p><a>Validate and set attribute value</a> <var>attr</var>'s <a for="Attr">value</a> for
+   <var>attr</var> with <var>element</var>.
+
+   <li><p>If <var>element</var> <a lt="has an attribute">has</a> an <a>attribute</a>
```
I've read through this spec path and I believe that, if the default policy removes the existing attribute node, then this will call replace an attribute, which in turn calls replace a list item, which results in a no-op because the old value no longer exists to be replaced.

So I think in this situation you're probably right the spec doesn't need to do anything, will make that change.

lukewarlow replied:
Having said that I'm slightly uneasy about stuff like attribute change steps still firing and how that all works spec wise.

Both Chromium and WebKit don't actually follow the spec 100% here and so I think would actually result in a different behaviour to this spec (the index lookup happens after the TT call) so that's also not ideal.

lukewarlow commented:
I think I've addressed all the comments from #1247.

I do want to point out Chrome and WebKit don't (or at least not in a way obvious to me) 100% follow the flow of the spec and as a result this may result in differences specifically in weird cases with attribute mutation.

So that bit especially it would be good to get feedback on.

It's also worth being aware that like Chromium's implementation this spec means that certain ways to update a nodes value don't work with a trusted type object as a user might expect. (e.g. `iframe.getAttributeNode('srcdoc').value = trustedHTMLObj;` will throw unless allowed by a default policy). I think in pratice this will be fine, and in some cases is the only real option we have without nodes keeping track of whether they're trusted or not (which would add lots of complexity)

annevk commented on dom.bs:
```diff
@@ -4279,7 +4281,7 @@ empty string instead, and then do as described below, switching on the interface
 <dl class=switch>
  <dt>{{Attr}}
- <dd><p><a>Set an existing attribute value</a> with <a>this</a> and the given value.
+ <dd><p><a>Set an existing attribute value</a> with <a>this</a>, the given value.
```
I don't understand this change.

lukewarlow replied:
apologies some of this is because stuff was added and then removed and I've clearly not reverted stuff very well

annevk commented on dom.bs:
```diff
@@ -6373,17 +6376,32 @@ given a <var>document</var>, <var>localName</var>, <var>namespace</var>, and opt
 
 <ol>
+ <li>
+  <p>If <var>validate</var>, then:
```
```suggestion
  <p>If <var>validate</var> is true, then:
```

annevk reviewed (commented):
This change is either incomplete or makes many cosmetic changes that would be best proposed separately as they confuse me quite a bit.

lukewarlow commented on dom.bs:
```diff
@@ -6875,11 +6915,14 @@ method steps are to return the result of
 
+<p>The <dfn method for=Element><code>setAttributeNodeNS(<var>attr</var>)</code></dfn> methods steps are to
+return the result of <a lt="set an attribute">setting an attribute</a> given <var>attr</var>,
+<a>this</a>.
```
Spotted this earlier, have fixed.

lukewarlow commented on dom.bs:
```diff
@@ -6422,6 +6440,18 @@ steps:
 
+<p>To <dfn id=concept-element-attributes-validate-and-set-value>validate and set attribute value</dfn>
+{{TrustedType}} or a string <var>value</var> for an <a>attribute</a> <var>attribute</var>, with
+<a for=/>element</a> <var>element</var>:
```
Changed this to Verify

lukewarlow commented:
> This change is either incomplete or makes many cosmetic changes that would be best proposed separately as they confuse me quite a bit.

Apologies I thought I'd reverted all of these changes but I missed a few, it's because I changed stuff and then changed it back and this led to some wonky diffs. Have hopefully reverted all of these unnecessary changes

annevk commented on dom.bs:
```diff
@@ -6373,17 +6376,32 @@ given a <var>document</var>, <var>localName</var>, <var>namespace</var>, and opt
+ <li><p>If <var>attribute</var>'s <a for=Attr>element</a> <a lt="has an attribute">has</a>
+ an <a>attribute</a> <var>attribute</var>, then <a>handle attribute changes</a> for
+ <var>attribute</var> with <var>attribute</var>'s <a for=Attr>element</a>, <var>oldValue</var>, and
+ <var>value</var>.
```
Should this be _value_ or _attribute_'s value? They can be different, right?

annevk commented on dom.bs:
```diff
@@ -6373,17 +6376,32 @@ given a <var>document</var>, <var>localName</var>, <var>namespace</var>, and opt
+ <var>attribute</var> with <var>attribute</var>'s <a for=Attr>element</a>, <var>oldValue</var>, and
+ <var>value</var>.
+
+ <li><p>Otherwise, throw an "{{InvalidStateError!!exception}}" {{DOMException}}.
```
Does this match the exception Trusted Types throws by default?

lukewarlow replied:
No Trusted Types throws a TypeError by default, not sure what's best to put here.

lukewarlow replied:
We no longer throw

annevk commented on dom.bs:
```diff
@@ -6373,17 +6376,32 @@ given a <var>document</var>, <var>localName</var>, <var>namespace</var>, and opt
- <li><p><a>Handle attribute changes</a> for <var>attribute</var> with <var>attribute</var>'s
- <a for=Attr>element</a>, <var>oldValue</var>, and <var>value</var>.
+ <li><p>If <var>attribute</var>'s <a for=Attr>element</a> <a lt="has an attribute">has</a>
+ an <a>attribute</a> <var>attribute</var>, then <a>handle attribute changes</a> for
```
"has an attribute" should be the linking text I think. Not sure why we'd mention the type as its own thing here.

annevk commented on dom.bs:
```diff
@@ -6373,17 +6376,32 @@ given a <var>document</var>, <var>localName</var>, <var>namespace</var>, and opt
 <p>To <dfn export id=concept-element-attributes-append lt="append an attribute">append</dfn> an
-<a>attribute</a> <var>attribute</var> to an <a for=/>element</a> <var>element</var>, run these
-steps:
+<a>attribute</a> <var>attribute</var> to an <a for=/>element</a> <var>element</var>, with optional
+boolean <var>validate</var> (default true), run these steps:
```
an optional boolean*

You can remove "run these steps" as well. We tend to leave that out these days when the algorithm is introduced with "To" and no return value is to be listed.

annevk commented on dom.bs:
```diff
@@ -6373,17 +6376,32 @@ given a <var>document</var>, <var>localName</var>, <var>namespace</var>, and opt
 <ol>
+ <li>
+  <p>If <var>validate</var> is true, then:
+  <ol>
```
Need a newline before `<ol>`.

annevk commented on dom.bs:
```diff
@@ -6373,17 +6376,32 @@ given a <var>document</var>, <var>localName</var>, <var>namespace</var>, and opt
 
 <ol>
+ <li>
+  <p>If <var>validate</var> is true, then:
```
```suggestion
  <p>If <var>validate</var> is true:
```

annevk commented on dom.bs:
```diff
@@ -6373,17 +6376,32 @@ given a <var>document</var>, <var>localName</var>, <var>namespace</var>, and opt
+  <p>If <var>validate</var> is true, then:
+  <ol>
+   <li><p><a>Verify and set attribute value</a>
+   <var>attribute</var>'s <a for="Attr">value</a> for <var>attribute</var> with <var>element</var>.
```
The text isn't consistent around comma usage when invoking this algorithm.

annevk commented on dom.bs:
```diff
@@ -6425,6 +6446,18 @@ steps:
+<a for=/>element</a> <var>element</var>:
+
+<ol>
+ <li><p>Let <var>validValue</var> be the result of calling
```
verifiedValue, presumably?

annevk commented on dom.bs:
```diff
@@ -6425,6 +6446,18 @@ steps:
  <a for=Attr>value</a>.
 </ol>
 
+<p>To <dfn id=concept-element-attributes-verify-and-set-value>verify and set attribute value</dfn>
```
Remove the id attribute here.

annevk commented on dom.bs:
```diff
@@ -6425,6 +6446,18 @@ steps:
 </ol>
 
+<p>To <dfn id=concept-element-attributes-verify-and-set-value>verify and set attribute value</dfn>
+{{TrustedType}} or string <var>value</var> for an <a>attribute</a> <var>attribute</var>, with
+<a for=/>element</a> <var>element</var>:
```
Perhaps:

> given a X x, Y y, and Z z:

as style instead?

lukewarlow replied:
I think that's done

annevk commented on dom.bs:
```diff
@@ -6485,8 +6518,6 @@ string <var>namespace</var> (default null):</p>
  <var>attr</var>'s <a for=Attr>namespace</a>, <var>attr</var>'s <a for=Attr>local name</a>, and
  <var>element</var>.
 
- <li><p>If <var>oldAttr</var> is <var>attr</var>, return <var>attr</var>.
```
This seems like it would change the mutation records for existing applications even when they don't use Trusted Types.

lukewarlow replied:
This doesn't seem intentional (likely a result of changes that have subsequently been partially reverted), I'll add it back

annevk commented:
Attributes are stored in a list and those do have indices per Infra. What am I missing?

lukewarlow commented:
Sorry I mean algorithmically the spec and implementations don't follow the same flow. So it's trickier to reason between the spec and implementation. This might just be my lack of familiarity with these APIs too.

annevk commented:
The path of least resistance is prolly matching Chromium. Introducing new paths that throw is always risky. If you are looking for guidance as to how, I'd need quite a bit more context to provide helpful suggestions.

lukewarlow commented:
> The path of least resistance is prolly matching Chromium. Introducing new paths that throw is always risky. If you are looking for guidance as to how, I'd need quite a bit more context to provide helpful suggestions.

I've partially matched Chromium but gone a bit further with the error correction as Chromium still has some bugs wrt mutating attributes within the default policy.

lukewarlow commented on dom.bs:
```diff
@@ -6530,23 +6545,39 @@ string <var>namespace</var> (default null):</p>
 
 <div algorithm>
 <p>To <dfn export id=concept-element-attributes-set-value>set an attribute value</dfn> given an
```
This algorithm covers setAttributeNS

lukewarlow commented on dom.bs:
```diff
@@ -6807,12 +6838,29 @@ method steps are:
- <a for=Attr>local name</a> is <var>qualifiedName</var>, <a for=Attr>value</a> is
- <var>value</var>, and <a for=Node>node document</a> is <a>this</a>'s <a for=Node>node document</a>,
- then <a lt="append an attribute">append</a> this <a>attribute</a> to <a>this</a>, and then return.
+ <li><p>Let <var>attributeExists</var> be false if <var>attribute</var> is null, and true otherwise.
```
This algorithm covers setAttribute

lukewarlow commented on dom.bs:
```diff
@@ -7382,7 +7430,13 @@ string <var>value</var>, run these steps:
  <a for=Attr>value</a> to <var>value</var>.
 
- <li><p>Otherwise, <a lt="change an attribute">change</a> <var>attribute</var> to <var>value</var>.
+ <li><p>Otherwise:
```
This covers attr.nodeValue, attr.value, and attr.textContent.

lukewarlow commented on dom.bs:
```diff
@@ -6519,6 +6529,11 @@ string <var>namespace</var> (default null):</p>
 
  <li><p>If <var>oldAttr</var> is <var>attr</var>, return <var>attr</var>.
 
+ <li><p>Let <var>verifiedValue</var> be the result of calling <a>verify attribute value</a>
```
This covers, setAttributeNode, setAttributeNodeNS, setNamedItem and setNamedItemNS

lukewarlow commented on dom.bs:
```diff
@@ -6870,11 +6918,17 @@ method steps are:
 
   <ol>
-   <li><p>If <var>force</var> is not given or is true, create an <a>attribute</a> whose
+   <li><p>Set <var>attribute</var> to an <a>attribute</a> whose
```
This covers toggleAttribute. I need to double check this because I think it could still lead to a weird state in it's current form.

lukewarlow replied:
Have updated this to work how chrome appears to

otherdaniel replied:
The TT-check in Chromium's `toggleAttribute` was unintentional. It came about as a side effect of how `toggleAttribute` was implemented on top of other methods. We can't come up with any XSS based on `toggleAttribute`. It'd probably be best to remove it from the PR, and that we fix our implementation on our end.

lukewarlow replied:
I've reverted this now so there's no enforcement in the spec. I will do a WebKit PR to remove the enforcement from there. And the existing chromium issue that discussed some of this can be used to remove it from Chromium.

annevk commented:
This now seems to invoke TT even when we have specifications manipulate attributes, or when `element.id`'s setter is used. Is that intentional?

For review it really helps to have an up-to-date commit message that clearly describes the intent. That's why we made it one of the PR checkboxes.

It also seems that it would cause a change in ordering between enforcing this in IDL and in the algorithms as now certain exceptions are thrown first. That will need test coverage.

lukewarlow commented:
> This now seems to invoke TT even when we have specifications manipulate attributes, 

I don't understand sorry could you explain what's changed since before?

> or when element.id's setter is used. Is that intentional?

That's not intentional? I don't see what I've changed that would lead that to happen?

In all cases if TT's algorithm is invoked it just returns the value you provide it, unless it's one of a select few atttributes on a select few elements (or an event handler).

I will write an overview first thing tomorrow and update the PR to provide that.

lukewarlow commented:
# Trusted Types integration with DOM

The DOM spec already contains partial integration with Trusted Types this was included in https://github.com/whatwg/dom/pull/1258

This PR updates the DOM functions related to attributes.

The following functions' IDL are adjusted, to accept a TrustedType (TrustedHTML or TrustedScript or TrustedScriptURL) or a string.

**Element.setAttribute**
**Element.setAttributeNS**

The above two functions are the "sanctioned" methods for updating the value of an attribute, any other way relies on default policy else they'll fail TT checks.

The algorithm for setAttribute function (and the "set an attribute value" algorithm used by setAttributeNS) is updated to call into the Trusted Types such that if the attribute being set is one of those covered by Trusted Types,
the value being set is verified (potentially being replaced with a different value).

These algorithms include some handling for if the trusted types default policy changes whether the attribute being set exists or not.

**Element.setAttributeNode** and **Element.setAttributeNodeNS** both, along with **NamedNodeMap.setNamedItem** and **NamedNodeMap.setNamedItemNS** make use of the "setting an attribute" algorithm, which is similarly updated to call into Trusted Types.

**Attr.value**, **Node.textContent**, and **Node.nodeValue** setters call "set an existing attribute value", this algorithm is updated such that for attributes attached to an element, trusted types is called into and verifies the value.

In all these cases the changes only have any impact for a select few attributes (e.g. event handlers, iframe srcdoc attribute)

lukewarlow commented:
The change to `set an attribute value` breaks iframe.srcdoc property setting currently (it'll do a double check).

smaug---- commented on dom.bs:
```diff
@@ -6519,6 +6529,11 @@ string <var>namespace</var> (default null):</p>
+ <var>attr</var>'s <a for=Attr>value</a> for <var>attr</var>, with <var>element</var>.
+
+ <li><p>Set <var>attr</var>'s <a for=Attr>value</a> to <var>verifiedValue</var>.
+
```
Hmm, so doing the verification may run scripts. And that means oldAttr might not be anymore in the element it used to be.   Could that cause issues? Could the value be validated first for certain kind of element but then used on some other kind of element?

lukewarlow replied:
The default policy doesn't provide context about which element an attribute is set on only the name of the attribute.  In this case this algorithm is triggered by APIs such as setAttributeNode or setNamedItem.

So I don't think there's anything that can happen here that's too bad. Also any mechanism you use inside of the default policy will itself trigger the default policy so it should be fine?

smaug---- replied:
The callback may have added the new attr to be in some other element's attribute list

fred-wang replied:
FWIW, in https://bugzilla.mozilla.org/show_bug.cgi?id=1944511 I'm adding new tests and will try to make sure this case and the other ones mentioned below are covered.

But my guess would be we should just throw the same exception as in step 1 in that case:

> 1 If attr’s [element](https://dom.spec.whatwg.org/#concept-attribute-element) is neither null nor element, [throw](https://webidl.spec.whatwg.org/#dfn-throw) an "[InUseAttributeError](https://webidl.spec.whatwg.org/#inuseattributeerror)" [DOMException](https://webidl.spec.whatwg.org/#idl-DOMException).

mozfreddyb replied:
The default policy only getting attribute information is an issue elsewhere, btw. https://github.com/w3c/trusted-types/issues/569 - Would be useful to add element information imho.

lukewarlow replied:
Good catch, I've added the fix suggested by Fred as that seems the most logical solution. I'll make sure to add a test against this when I fix it in WebKit.

fred-wang replied:
Just for the record, this is https://phabricator.services.mozilla.com/D236107?id=977466#inline-1311431 and I agree with @smaug---- this is a problem in my current patch. I also started to add test for this and other cases in https://phabricator.services.mozilla.com/D236161#change-XQGcauHgW6CO

lukewarlow replied:
This is fixed and has a test, per Daniels' suggestion and my recent spec change this should be even cleaner. It will require a test change to verify the order of exceptions (the TT TypeError / default policy exception is now the first exception thrown)

smaug---- commented on dom.bs:
```diff
@@ -7382,7 +7437,13 @@ string <var>value</var>, run these steps:
+    <li><p>Let <var>verifiedValue</var> be the result of calling <a>verify attribute value</a>
+    <var>value</var> for <var>attribute</var>, with <a>this</a>.
+
+   <li><p><a lt="change an attribute">Change</a> <var>attribute</var> to <var>verifiedValue</var>.
```
Doesn't this lead to effectively null pointer crashes in algorithms if attribute has been moved out from an element by the tt callback?

lukewarlow replied:
I guess potentially if some sub algorithm of change an attribute is assuming that element must exist? Fwiw in practice this doesn't cause issues from what I can see.

Not sure what the best fix here would be?

fred-wang replied:
I haven't had time to do much progress since my previous comments here, so this is mostly guesses based on the change to dom/base/Attr.cpp from https://phabricator.services.mozilla.com/D236107. Mozilla's "change the attribute" is indeed calling a SetAttr method on the element and I see GetTrustedTypesCompliantAttributeValue could:

1. Detach the attribute node: in that the attribute element's is null and probably we should change the spec to do a no-op like in "If attribute's element is null, then set attribute's value to value" above.
2. Destroy the original element: this means implementations need to hold a strong reference to the element or skip the SetAttr call. However, I believe the attribute somehow keeps a strong reference to the element it is attached to, so this is probably covered by 1 already.
3. Move the attribute to another element. Again, I believe this will effectively detached the attribute node from the original element, so covered by 1 already.

fred-wang replied:
Sorry, not a "no-op" I mean when the attribute's element is null then Mozilla does not need to call Element:SetAttr, just to set Attr::mValue directly. So probably we don't need a spec change here, but implementations need to properly deal with the different cases mentioned and we need tests for them.

fred-wang replied:
I see Chromium / WebKit just call their element's "set attribute", which takes care of applying TT policy default:

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/attr.cc;drc=003b43cb850a2fb81bb482cd95381338b7205933;l=73
https://searchfox.org/wubkat/rev/446d00e44d23eba2d8c1c1c92bf0171472dee385/Source/WebCore/dom/Attr.cpp#97

They both store the element in a strong reference (RefPtr and Member) so it won't get destroyed by the callback. And they will modify the attribute value of that element.

But the spec does not check whether the attribute's element become null. And if it is attached it to a different element, then it will just run the "handle attribute changes" on that new element:
https://dom.spec.whatwg.org/#concept-element-attributes-change
https://dom.spec.whatwg.org/#handle-attribute-changes

lukewarlow replied:
Having discussed this with Fred we've got an idea on what to do here. The spec will say that if the element is changed by the TT callback that we early return (perhaps throw in which case we need to decide an exception).

smaug---- reviewed (commented):
Some questions. I could be missing something in TT spec which somehow covers the problematic cases, but please at least explain and ask review again.

lukewarlow commented on dom.bs:
```diff
@@ -6457,6 +6457,16 @@ steps:
+<ol>
+ <li><p>Return the result of calling
+ <a abstract-op>get Trusted Types-compliant attribute value</a> for <var>attribute</var>, with
+ <var>element</var>, <var>value</var>. [[!TRUSTED-TYPES]]
```
They will end up rethrowing the exception which is what's expected.

annevk commented:
Now that 809bfa24535b1d55d761d8add353e1c3fde05801 is merged this should add the Infrastructure changes back in again as this now introduces the dependency on Trusted Types.

domfarolino commented on dom.bs:
```diff
@@ -7435,7 +7492,13 @@ string <var>value</var>, run these steps:
+    <li><p>Let <var>verifiedValue</var> be the result of calling <a>verify attribute value</a>
+    <var>value</var> for <var>attribute</var>, with <a>this</a>.
+
+   <li><p><a lt="change an attribute">Change</a> <var>attribute</var> to <var>verifiedValue</var>.
```
Indenting of this line is wrong. Either that, or the first `<li>` above is indented wrong.

lukewarlow replied:
I've ran specfmt on this PR, hopefully that address this?

lukewarlow replied:
Apologies Idk if this specfmt got lost by a rebase but I think things have been corrected by my recent change.

domfarolino reviewed (commented):
So I came across this in my inbox, very belatedly, but the PR seems to have been in a lull for a bit. What is the current status? Seems there are a few comment threads that are open and unresolved, but no major blockers identified I guess? Let me know if I can help.

mbrodesser-Igalia commented:
https://github.com/w3c/trusted-types/issues/438 was reverted and additionally mentions `ChildNode`'s `before`, `after` and `replaceWith`.

https://www.w3.org/TR/trusted-types/#integration-with-dom refers to this PR. So should this PR cover `ChildNode`'s methods too?

CC @fred-wang

mbrodesser-Igalia commented:
https://www.w3.org/TR/trusted-types/#integration-with-dom links to this ticket.
So this ticket should either cover the APIs mentioned at https://github.com/w3c/trusted-types/issues/438 too, or they should be dealt with in a separate ticket.

CC @fred-wang

fred-wang commented:
>  Tests are written and can be reviewed and commented upon at:
   https://github.com/web-platform-tests/wpt/tree/master/trusted-types

On https://github.com/whatwg/dom/pull/1268#issuecomment-2165953464 you mentioned the affected APIs are

- setAttribute
- setAttributeNS
- Element.setAttributeNode
- Element.setAttributeNodeNS
- NamedNodeMap.setNamedItem
- NamedNodeMap.setNamedItemNS
- Attr.value
- Node.textContent
- Node.nodeValue

Is that an exhautive list?

I think we would need tests for each of the affected API to check:
- That attribute change is blocked by the default policy.
- That a non-trivial default policy is applied before attribute change.

I found the following tests:
- block-string-assignment-to-attribute-via-attribute-node.html
- block-string-assignment-to-Element-setAttribute.html
- testing/web-platform/tests/trusted-types/TrustedType-AttributeNodes.html

But they don't cover e.g. setAttributeNodeNS or setNamedItemNS and the last two are a bit messy (https://phabricator.services.mozilla.com/D227943 was enough to make the last one pass in Firefox, which seems dubious to me).

otherdaniel commented:
Hi all, sorry I'm a bit late with feedback. It turns out, what Chromium does here is buggy since we are running structure checks before the Trusted Type checks and then getting inconsistent results when a default policy modifies the attribute that is being checked. Thanks for pointing this out.

My (new) preference would be to always run TT checks before any other steps; here and elsewhere. We mostly already do this. That would provide a strict sequencing of TT before any other DOM checks or DOM manipulations, and should IMHO solve this class of problems entirely. I'd be happier if we had a general rule here, and I'd rather do some more work on our end than risk us overlooking another issue like this.

For testing, I'd propose to add WPT tests with competing error conditions and checking for the correct exception being thrown. (I.e., setting an attribute node that's already connected should throw the TT-style TypeException, not the InUseAttributeError DOMException). Something like the following seems to work when I try it locally:

```js
<div onclick="1+1;" id="test"></div>
<script>
test(t => {
  var attribute = document.getElementById("test").getAttributeNode("id");
  assert_throws_dom("InUseAttributeError", _ => {
    document.body.setAttributeNode(attribute);
  }, "Expect InUseAttributeError DOMException.");

  attribute = document.getElementById("test").getAttributeNode("onclick");
  assert_throws_js(TypeError, _ => {
    document.body.setAttributeNode(attribute);
  }, "Expect a Trusted-Types style TypeError, not InUseAttributeError DOMException.");
}, "Element.setAttribute competing exceptions");
</script>
```

fred-wang commented:
Just to be more explicit, I believe we have three categories of APIs:

- Element.setAttribute/Element.setAttributeNS: It seems browsers agree to always successfully set the attribute in any case. If the attribute node is moved to another element in the default's policy callback, we end up with the attribute set on two different elements.

- Element.setAttributeNode, Element.setAttributeNodeNS, NamedNodeMap.setNamedItem, NamedNodeMap.setNamedItemNS: If the attribute node is moved to another element in the default's policy callback, WebKit and Chromium may end up associating the attribute node to two different element's attribute maps. My proposal is to throw an InUseAttributeError error if the attribute node was moved to another element (which is what would be happening if we re-run https://dom.spec.whatwg.org/#concept-element-attributes-set without TT check).

- Attr.value, Attr.nodeValue, Attr.textContent: These APIs essentially operate on an attribute. If the attribute's element is non-null, then WebKit and Chromium just calls setAttributeNS on that element and no new issue arises. The spec instead performs the TT check before setting the element's attribute but does not say what to do when the attribute's element changes (i.e. attribute is detached from the element or moved to another element). My proposal is again to just re-run current https://dom.spec.whatwg.org/#set-an-existing-attribute-value without TT check. This would differ from WebKit/Chromium's current implementation.

I updated https://phabricator.services.mozilla.com/D236161 to match this proposal ; and https://phabricator.services.mozilla.com/D236107 makes all these new tests pass. I'm happy to change to something else depending on what we decide at the end.

otherdaniel commented:
The solution outlined here is correct & self-consistent, and if all implementations agree I'll be happy.

I still think we can be simpler, though: The default policy introduces a seemingly concurrent operation. All operations, on all three categories of APIs outlined above, can be explained by serializing these operations.

Consider setAttributeNode:

As spec'ed here: First check structure; then run TT check; then re-check structure.
    - TT error => throws TypeException
    - structure error but no TT error => DOMException AttributeInUseError
    - structure error created in TT callback => DOMException AttributeInUseError
    - TT error *and* structure error =>TypeException
 
If instead, we spec it like so:
1. Let verifiedValue be value.
2. If verify is true: [... TT check; assign to verifiedValue ...]
3. If attr’s element is neither null nor element, throw an "InUseAttributeError" DOMException
4. [... old spec, but verifiedValue instead of value ...]

I think this would get us all of the same error conditions, but fewer operations. And IMHO easier to understand. And I like it, because explaining this behaviour in terms of serializing the checks and running the TT check gives us a guideline which we can apply to all similar cases.

There is one observable difference I can think of: If we have a structure error, and if the TT policy itself throws an exception, then in the first case we'd get AttributeInUseError, and in the second we'd get TypeException. I think that's okay.

annevk commented:
Any updates here? I rather like @otherdaniel's suggestion.

lukewarlow commented:
If others are happy with that idea I'm happy to look into updating this PR to match.

smaug---- commented:
I think that would work for setNameItem* and setAttributeNode*.
Current code in Gecko does "If attr’s element is neither null nor element" before and after, and maybe doing it only after is good enough. right @fred-wang ?

Attr.value/.nodeValue/.textContent should do validation (if there is element) too before modifying the value.
(chromium and webkit seem to do something else, assuming I'm reading the code correctly)

Element.setAttribute/Element.setAttributeNS  in the PR needs some tweaking too. Implementations don't necessarily create Attr nodes when those APIs are used, but if they do, what happens if TT callback moves the Attr away. As fred-wang mentioned earlier, implementations already just create a new attribute on the element.

annevk commented on dom.bs:
```diff
@@ -6589,6 +6591,14 @@ steps:
+
+<ol>
+ <li><p>Return the result of calling <a abstract-op>get Trusted Types-compliant attribute value</a>
+ for <var>attribute</var>, with <var>element</var>, <var>value</var>. [[!TRUSTED-TYPES]]
```
```suggestion
 for <var>attribute</var>, with <var>element</var> and <var>value</var>. [[!TRUSTED-TYPES]]
```
Although maybe we can use the more standard "given attribute, element, and value"?

annevk commented on dom.bs:
```diff
@@ -6641,6 +6651,9 @@ string <var>namespace</var> (default null):</p>
 <a for=/>attribute</a> <var>attr</var> and an <a for=/>element</a> <var>element</var>:
 
 <ol>
+ <li><p>Let <var>verifiedValue</var> be the result of calling <a>verify attribute value</a>
+ <var>attr</var>'s <a for=Attr>value</a> for <var>attr</var>, with <var>element</var>.
```
```suggestion
 <li><p>Let <var>verifiedValue</var> be the result of <a>verifying attribute value</a> given
 <var>attr</var>'s <a for=Attr>value</a>, <var>attr</var>, and <var>element</var>.
```
However, this doesn't match the signature above. What gives?

lukewarlow replied:
I've updated this and the signature to hopefully be clearer. I think it does match the signature it was just written rather poorly. I've changed the signature to a more convetional one and updated the call sites accordingly.

annevk commented on dom.bs:
```diff
@@ -6656,29 +6669,53 @@ string <var>namespace</var> (default null):</p>
- <a for=Node>node document</a> is <var>element</var>'s <a for=Node>node document</a>, then
- <a lt="append an attribute">append</a> this <a>attribute</a> to <var>element</var>, and then
- return.
+ <li><p>Let <var>attributeExists</var> be false if <var>attribute</var> is null, and true otherwise.
```
```suggestion
 <li><p>Let <var>attributeExists</var> be false if <var>attribute</var> is null; otherwise true.
```

annevk commented on dom.bs:
```diff
@@ -6656,29 +6669,53 @@ string <var>namespace</var> (default null):</p>
+
+ <li><p>Let <var>verifiedValue</var> be <var>value</var>.
+
+ <li><p>If <var>verify</var> is true:
```
`<p>` on its own line if the `<li>` has multiple children.

annevk commented on dom.bs:
```diff
@@ -6656,29 +6669,53 @@ string <var>namespace</var> (default null):</p>
+ <li><p>If <var>attributeExists</var> is true, <a lt="change an attribute">change</a>
+ <var>attribute</var> to <var>verifiedValue</var>.
+
+ <li><p>Otherwise:
```
`<p>` on its own line as `<li>` has multiple children.

annevk commented on dom.bs:
```diff
@@ -6656,29 +6669,53 @@ string <var>namespace</var> (default null):</p>
+   <li><p>Set <var>verifiedValue</var> to the result of calling <a>verify attribute value</a>
+   <var>value</var> for <var>attribute</var>, with <var>element</var>.
+
+   <li><p>Set <var>attributeExists</var> to true if <var>element</var> <a lt="has an attribute">has
```
No newlines in phrasing level elements.

lukewarlow replied:
I think I've addressed all of these, will check the github diff to make sure

annevk commented on dom.bs:
```diff
@@ -6939,12 +6976,27 @@ method steps are:
- <a for=Attr>local name</a> is <var>qualifiedName</var>, <a for=Attr>value</a> is
- <var>value</var>, and <a for=Node>node document</a> is <a>this</a>'s <a for=Node>node document</a>,
- then <a lt="append an attribute">append</a> this <a>attribute</a> to <a>this</a>, and then return.
+ <li><p>Let <var>attributeExists</var> be false if <var>attribute</var> is null, and true otherwise.
```
```suggestion
 <li><p>Let <var>attributeExists</var> be false if <var>attribute</var> is null; otherwise true.
```

annevk commented on dom.bs:
```diff
@@ -7514,7 +7566,13 @@ string <var>value</var>, run these steps:
- <li><p>Otherwise, <a lt="change an attribute">change</a> <var>attribute</var> to <var>value</var>.
+ <li><p>Otherwise:
+  <ol>
+    <li><p>Let <var>verifiedValue</var> be the result of calling <a>verify attribute value</a>
```
Single-space indentation.

annevk reviewed (commented):
I did a quick editorial review. Please double check the comments throughout as some apply several times.

domfarolino commented on dom.bs:
```diff
@@ -6656,29 +6669,57 @@ string <var>namespace</var> (default null):</p>
-<a for=/>element</a> <var>element</var>, a string <var>localName</var>, a string <var>value</var>,
-an optional null or string <var>prefix</var> (default null), and an optional null or string
-<var>namespace</var> (default null):
+<p>To <dfn export id=concept-element-attributes-set-value>set an attribute value</dfn> given an <a
```
This was probably done by `specfmt`, as it basically follows the HTML wrapping rules where this is allowed. Would you prefer we not use it on non-HTML specs for now? I recall you being sympathetic to allowing this kind of wrapping outside of HTML, but maybe you'd prefer to do that only if we update the entire spec to that model.

annevk replied:
I think I said that I only want it to be used once we know it can be used in automation and then I want it enforced by automation. Perhaps I've been inconsistent though.

lukewarlow replied:
I'll stop using it for Dom and fix these up manually then.

otherdaniel commented on dom.bs:
```diff
@@ -7514,7 +7572,23 @@ string <var>value</var>, run these steps:
+   given <var>value</var>, <var>attribute</var>, and <a>this</a>.
+
+   <li><p>If <var>attribute</var>'s <a for=Attr>element</a> is null, then set <var>attribute</var>'s
+   <a for=Attr>value</a> to <var>value</var>, and return.
```
Should this set Attr/value to `verifiedValue` (instead of `value`)?

WPT trusted-types/set-attributes-mutations-in-callback.tentative.html seems to expect verifiedValue here, and I get a number test failures if I implement this as written.

lukewarlow replied:
Cc @fred-wang can you remember what we discussed. I seem to recall that maybe the tests for this specific function still need some work?

verifiedValue would make sense on the surface but there's also maybe a reason why it doesn't.

lukewarlow replied:
(Apologies for the delay)

I just looked into this and I think it should indeed be verifiedValue. I'll update the spec PR now. Just as an fyi that test specifically is out of date with the spec currently so you'll probably still gets some failures.

smaug---- commented on dom.bs:
```diff
@@ -6939,12 +6982,30 @@ method steps are:
+ <var>attribute</var>, and <a>this</a>.
+
+ <li><p>Set <var>attributeExists</var> to true if <a>this</a>
+ <a lt="has an attribute">has an attribute</a> <var>attribute</var>; otherwise false.
```
As I said on Matrix, I wish we wouldn't need to do identity checks on attributes. That would be a bit tricky, or at least slow in Gecko, and I think also in other engines.

lukewarlow replied:
per matrix discussion I've updated this and the setAttributeNS equivalent.

smaug---- commented on dom.bs:
```diff
@@ -6667,18 +6683,32 @@ an optional null or string <var>prefix</var> (default null), and an optional nul
+
+ <li><p>Let <var>verifiedValue</var> be the result of
+ <a lt="verify attribute value">verifying an attribute value</a> given <var>value</var>,
+ <var>attribute</var>, and <var>element</var>.
```
Hmm, why do we need this algorithm to verify the value? I'd expect the relevant webidl attribute setters to be modified, similarly to set HTMLScriptElement.src.

In fact, HTMLIFrameElement.srcdoc setter already does TT check to get a reasonable string and this would do it again? Or am I missing something.

lukewarlow replied:
But this is for setAttributeNS()? Does this also get used by property setters?

lukewarlow replied:
Ah yeah you're right this is shared by element reflection...

So yeah this will need moving out to the method algorithm. I think in this case it's probably different enough to just duplicate it a bit.

annevk commented on dom.bs:
```diff
@@ -6946,6 +6948,14 @@ steps:
  <a for=Attr>value</a>.
 </ol>
 
+<p>To <dfn>verify attribute value</dfn> given a {{TrustedType}} or string <var>value</var>, an
```
```suggestion
<p>To <dfn>verify an attribute value</dfn> given a {{TrustedType}} or string <var>value</var>, an
```
Although why do we even have this algorithm as it directly calls into the Trusted Types algorithm.

annevk commented on dom.bs:
```diff
@@ -7294,14 +7310,26 @@ method steps are:
  and null otherwise.
- <!-- This is step 2 of "get an attribute by name", modified as appropriate -->
+
+ <li><p>If <var>attribute</var> is null, then set <var>attribute</var> to an <a>attribute</a>
+ whose <a for=Attr>local name</a> is <var>qualifiedName</var>, <a for=Attr>value</a> is
+ <var>value</var>, and <a for=Node>node document</a> is <a>this</a>'s <a for=Node>node document</a>.
```
Is the idea here to create a dummy attribute that you don't append in order to calculate the verified value?

Why is that done this way? This is quite confusing to read through.

lukewarlow replied:
Apologies for the slow reply on here I thought I replied before but it must have got lost.

TLDR: I agree with you I can't really remember why it's done this way. I updated the TT spec to just take the local name and namespace directly rather than an attribute object. As part of doing this change I realised these algorithms were overly complicated for what was actually needed.

Hopefully now the integration points should be much cleaner and less invasive.

annevk reviewed (commented):
I would love to have a more detailed description of what we are trying to accomplish here. I don't think the resulting algorithms are very clear. Have you looked at whether it's possible to abstract certain bits?

lukewarlow commented:
@otherdaniel I believe this spec PR fully matches your proposals to call TT as early as possible now. If you've got time I'd be thankful for a review to get another pair of eyes on this.

lukewarlow commented on dom.bs:
```diff
@@ -7013,6 +7020,8 @@ string <var>namespace</var> (default null):</p>
 
  <li><p>Otherwise, <a lt="append an attribute">append</a> <var>attr</var> to <var>element</var>.
 
+ <li><p>Set <var>attr</var>'s <a for=Attr>value</a> to <var>verifiedValue</var>.
```
So I double checked and for point 1 there's actually an interop bug here, WebKit (and Gecko) doesn't throw a TT error. Chromium currently does, so we can choose whichever behaviour we think is best here.

cc @annevk  and @otherdaniel for your opinions

TLDR should doing

`element.setAttributeNode(element.attributes[1])` - throw a TT error when TT is enforced (and where element.attributes[1] is something like an onclick attribute?

annevk replied:
I think we either have to throw or do nothing at all, right? The latter is very weird as even no-op mutations result in a mutation record being queued. And queuing a mutation record for trusted type attributes without validation having occurred seems bad, so I think we have to throw.

annevk reviewed (commented):
This is indeed a whole bit more straightforward to read through.

lukewarlow commented on dom.bs:
```diff
@@ -7891,7 +7908,25 @@ string <var>value</var>, run these steps:
+   <a for=Attr>local name</a>, <var>attribute</var>'s <a for=Attr>namespace</a>, <a>this</a>,
+   and <var>value</var>. [[!TRUSTED-TYPES]]
+
+   <li><p>If <var>attribute</var>'s <a for=Attr>element</a> is null, then set <var>attribute</var>'s
+   <a for=Attr>value</a> to <var>verifiedValue</var>, and return.
+
+   <li><p>If <var>attribute</var>'s <a for=Attr>element</a> is not <var>originalElement</var>, then
+   return.
```
Regarding does this get called early enough.

"set an existing attribute value" is the first step invoked by Attr's value setter steps, Attr's textContent setter steps, and Attr's nodeValue setter steps.

NamedNodeMap's setNamedItem and setNamedItemNS, along with Element's setAttributeNode and setAttributeNodeNS both call "set an attribute" as their first steps.

Then setAttribute and setAttributeNs both have their own dedicated steps.

annevk commented on dom.bs:
```diff
@@ -7115,6 +7117,11 @@ string <var>namespace</var> (default null):</p>
 <a for=/>attribute</a> <var>attr</var> and an <a for=/>element</a> <var>element</var>:
 
 <ol>
+ <li><p>Let <var>verifiedValue</var> be the result of calling <a abstract-op>get
+ Trusted Types-compliant attribute value</a> with <var>attr</var>'s <a for=Attr>local name</a>,
```
I don't understand why this is an abstract-op. Generally outside of TC39 we should not be using that.

The wrapping is also wrong for DOM (applies several times). There's no wrapping inside of phrasing-level elements to make searching easier.

annevk reviewed (commented):
@fred-wang @otherdaniel if you could reply to @lukewarlow's comment above that would help a lot, thanks! @smaug---- did you want to have another look as well?

smaug---- commented on dom.bs:
```diff
@@ -8028,7 +8045,25 @@ string <var>value</var>, run these steps:
+   <a for=Attr>value</a> to <var>verifiedValue</var>, and return.
+
+   <li><p>If <var>attribute</var>'s <a for=Attr>element</a> is not <var>element</var>, then
+   return.
```
So, I would drop this check. no-op isn't good, and if TT callback has moved the attribute, then it has. We need to trust TT callbacks in that sense.

annevk commented:
Indeed, thanks a lot @lukewarlow for finishing this and @smaug---- for his careful reviews! And of course everyone else that helped improve Trusted Types integration over the years. 🎉
